pkg/gadgets/ebpftop/piditer/iter.go

```diff
@@ -0,0 +1,190 @@
+	bpfProgFopsAddr uint64
+	triedGetAddr    bool
+	cspecLock       sync.Mutex
+	triedGetSpec    bool
+	cspec           *ebpf.CollectionSpec
+	cspecErr        error
+)
+
+func getSpec() (*ebpf.CollectionSpec, error) {
+	cspecLock.Lock()
+	defer cspecLock.Unlock()
+
+	if triedGetSpec {
+		return cspec, cspecErr
+	}
+
+	triedGetSpec = true
+
+	tmpSpec, err := loadPiditer()
+	if err != nil {
+		cspecErr = err
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	consts := map[string]interface{}{
+		"bpf_prog_fops_addr": getBpfProgFopsAddr(),
+	}
+
+	if err := tmpSpec.RewriteConstants(consts); err != nil {
+		cspecErr = err
+		return nil, fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	cspec = tmpSpec
+
+	return cspec, nil
+}
+
+// getBpfProgFopsAddr reads the address of bpf_prog_fops from /proc/kallsyms. It
+// only does this on first call and caches the address.
+func getBpfProgFopsAddr() uint64 {
+	addrLock.Lock()
+	defer addrLock.Unlock()
+
+	if triedGetAddr {
+		return bpfProgFopsAddr
+	}
+
+	triedGetAddr = true
+
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		if fields[2] == "bpf_prog_fops" {
+			addr, err := strconv.ParseUint(fields[0], 16, 64)
+			if err != nil {
+				return 0
+			}
+
+			// cache addr
+			bpfProgFopsAddr = addr
+			return bpfProgFopsAddr
+		}
+	}
+	return bpfProgFopsAddr
+}
+
+func (p *pidIter) iteratePids() ([]*PidIterEntry, error) {
+	spec, err := getSpec()
+	if err != nil {
+		return nil, err
+	}
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&p.objs, &opts); err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	iter, err := link.AttachIter(link.IterOptions{
+		Program: p.objs.Iter,
+		Map:     nil,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach iter: %w", err)
+	}
+	defer iter.Close()
+
+	rc, err := iter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("failed to open iter: %w", err)
+	}
+
+	defer rc.Close()
+	buf := make([]byte, 4096/24*24)
+
+	res := make([]*PidIterEntry, 0)
+	for {
+		n, err := io.ReadFull(rc, buf)
+		if err != nil {
+			if errors.Is(err, unix.EAGAIN) {
+				continue
+			}
+		}
+		if n == 0 {
+			break
+		}
+		if n%24 != 0 {
+			return nil, fmt.Errorf("invalid format: %d", n)
+		}
+		for i := 0; i < n/24; i++ {
+			entry := (*C.struct_pid_iter_entry)(unsafe.Pointer(&buf[i*24]))
+			res = append(res, &PidIterEntry{
+				ProgID: uint32(entry.id),
+				Pid:    uint32(entry.pid),
+				Comm:   C.GoString(&entry.comm[0]),
+			})
+		}
+	}
+
+	return res, nil
+}
+
+// IteratePids returns an array of PidIterEntry containing information
+// on which pid (and comm) has an open fd to which eBPF Program ID.
+func IteratePids() ([]*PidIterEntry, error) {
+	return (&pidIter{}).iteratePids()
```

Comment:
With this, the optimisation with `triedGetAddr` and `triedGetSpec` become unnecessary because this code would not be executed every second but only when the trace is created.